File: Ahorn/entities/PredatorPlant.jl
module JungleHelperPredatorPlant

using ..Ahorn, Maple

@mapdef Entity "JungleHelper/PredatorPlant" PredatorPlant(x::Integer, y::Integer, color::String="Pink", facingRight::Bool=false)

const colors = String["Pink", "Blue", "Yellow"]

const placements = Ahorn.PlacementDict(
    "Predator Plant ($(color)) (Jungle Helper)" => Ahorn.EntityPlacement(
        PredatorPlant,
        "point",
        Dict{String, Any}(
          "color" => color
        )
    ) for color in colors
)

const spritePaths = Dict{String,String}(
    "Pink" => "JungleHelper/PredatorPlant/Bitey00",
    "Blue" => "JungleHelper/PredatorPlant/Blue/BiteyB00",
    "Yellow" => "JungleHelper/PredatorPlant/Yellow/BiteyC00"
)

Ahorn.editingOptions(entity::PredatorPlant) = Dict{String, Any}(
    "color" => colors
)

function Ahorn.selection(entity::PredatorPlant)
    x, y = Ahorn.position(entity)
    facingRight = get(entity.data, "facingRight", false)
    return Ahorn.Rectangle(x - (facingRight ? 18 : 6), y - 16, 24, 24)
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::PredatorPlant, room::Maple.Room)
    color = get(entity.data, "color", "Pink")
    scaleX = get(entity.data, "facingRight", false) ? -1 : 1
    Ahorn.drawSprite(ctx, spritePaths[color], 0, -4, sx=scaleX)
end

end

File: Loenn/entities/predatorPlant.lua
local utils = require("utils")

local predatorPlant = {}
predatorPlant.name = "JungleHelper/PredatorPlant"

local colors = { "Pink", "Blue", "Yellow" }

local spritePaths = {
    Pink = "JungleHelper/PredatorPlant/Bitey00",
    Blue = "JungleHelper/PredatorPlant/Blue/BiteyB00",
    Yellow = "JungleHelper/PredatorPlant/Yellow/BiteyC00"
}

predatorPlant.placements = {}

for _, color in ipairs(colors) do
    table.insert(predatorPlant.placements, {
        name = color,
        data = {
            color = color,
            facingRight = false,
            sprite = "",
        }
    })
end

predatorPlant.fieldInformation = {
    color = {
        options = colors,
        editable = false
    }
}

predatorPlant.offset = { 16, 16 }

function predatorPlant.texture(room, entity)
    return spritePaths[entity.color]
end

function predatorPlant.scale(room, entity)
    return entity.facingRight and -1 or 1, 1
end

function predatorPlant.selection(room, entity)
    return utils.rectangle(entity.x - (entity.facingRight and 18 or 6), entity.y - 16, 24, 24)
end

return predatorPlant
